### PR TITLE
BC: Use Torch DataLoader

### PR DESCRIPTION
**Pull request by shwang**
Make `BC` accept a Torch `DataLoader`-like type (`Iterable[Dict[str, Tensor]]`) rather than our custom `Dataset`.

This is PR is the first in a series of two that will phase our custom `Dataset` abstract class. The next PR will phase out `Dataset` entirely by using Torch data objects in `Adversarial`.

One consequence of using `DataLoader` instead of our `Dataset`, which has the `.sample(batch_size)` abstract method, is that `BC` can no longer controls the batch_size. Instead the `DataLoader` set the batch size implicitly in the arbitrary batches yielded from its `__iter__` method. 

To make `Transitions` passable into `DataLoader.__init__` the `torch.util.Dataset` argument, I implemented `Transitions.__getitem__` method so that `Transitions` can subclass `torch.util.data.Dataset`. Only indexing was required to subclass,  but I also implemented slicing. To avoid unexpected `ValueError` in `Transitions`-slicing from length-zero slices, I removed the `Transitions.__postinit__` check that bans zero-length instances of `Transitions`.

### CODE REVIEW

File: src/imitation/algorithms/dagger.py [hunk 1]
```diff
@@ -310,7 +314,10 @@ def _try_load_demos(self):
             logging.info(
                 f"Loaded {sum(num_demos)} new demos from {len(num_demos)} rounds"
             )
-            self.bc_trainer.set_expert_dataset(transitions)
+            dataloader = th_data.DataLoader(
+                transitions, self.batch_size, drop_last=True
+            )
```
Comment (shwang) [inline src/imitation/algorithms/dagger.py:319]: We could make this more customizable beyond `batch_size` but this seems like a reasonable default to start with. (Equivalent to behavior before this commit, except without shuffling.)

- [ ] Add `shuffle=True`.

**Comment by AdamGleave** [comment]
This means we can no longer (easily) have algorithms with different batch sizes over time? Seems OK, a rather esoteric use case, but something to be aware of at least.

File: tests/test_bc.py [hunk 2]
```diff
@@ -19,24 +20,49 @@ def venv():
...
+def trainer(batch_size, venv, use_ducktyped_dataloader):
     rollouts = types.load(ROLLOUT_PATH)
-    data = rollout.flatten_trajectories(rollouts)
-    if convert_dataset:
-        data = datasets.TransitionsDictDatasetAdaptor(
-            data, datasets.EpochOrderDictDataset
-        )
-    return bc.BC(venv.observation_space, venv.action_space, expert_data=data)
+    trans = rollout.flatten_trajectories(rollouts)
+    if not use_ducktyped_dataloader:
```
Comment (AdamGleave) [inline tests/test_bc.py:51]: Nit: it feels more natural to me to write `if use_ducktyped_dataloader` and swap the two branches.

File: tests/test_data.py [hunk 3]
```diff
@@ -146,11 +148,42 @@ def test_valid_transitions(
...
+                start = np.random.randint(-2, length)
+                stop = np.random.randint(0, length + 2)
+                step = np.random.randint(-2, 4)
+                if step == 0:  # Illegal. Quick fix that biases tests to ordinary step.
+                    step = 1
+                s = slice(start, stop, step)
+                assert type(trans[s]) is type(trans)
+                self._check_transitions_get_item(trans, s)
+
+    def _check_transitions_get_item(self, trans, key):
```
Comment (AdamGleave) [inline tests/test_data.py:174]: Why is this an instance method? It does not depend on `self` as far as I can see, so it could be a module level function instead?

File: tests/test_data.py [hunk 4]
```diff
@@ -146,11 +148,42 @@ def test_valid_transitions(
...
+                if length != 0:
+                    index = np.random.randint(length)
+                    assert isinstance(trans[index], dict)
+                    self._check_transitions_get_item(trans, index)
+
+                # Slicing checks.
+                start = np.random.randint(-2, length)
+                stop = np.random.randint(0, length + 2)
+                step = np.random.randint(-2, 4)
+                if step == 0:  # Illegal. Quick fix that biases tests to ordinary step.
```
Comment (AdamGleave) [inline tests/test_data.py:168]: This seems fine, but if you wanted to avoid bias you could do `if step >= 0: step += 1`, and then it would be randomly sampling from the uniform interval excluding 1.

File: src/imitation/algorithms/bc.py [hunk 5]
```diff
@@ -50,6 +49,9 @@ def __call__(self, _):
         return self.lr
 
 
+BCDataLoaderLike = Iterable[Mapping[str, th.Tensor]]
```
[the remark this answers is not in the document]
Comment (shwang) [reply src/imitation/algorithms/bc.py:52]: The hyphen would go "DataLoader-like". Basically anything ducktypes a `torch.util.data.DataLoader` returning the correct data for BC.

I'll think of a better name for this type alias.

File: src/imitation/data/types.py [hunk 6]
```diff
@@ -109,15 +119,39 @@ def __post_init__(self):
...
+        d_item = {k: v[key] for k, v in d.items()}
+
+        if isinstance(key, slice):
+            # Return type is the same as this dataclass. Replace field value with
+            # slices.
+            return dataclasses.replace(self, **d_item)
+        else:
+            assert isinstance(key, int)
+            # Return type is a dictionary. Array values have no batch dimension.
+            return d_item
```
[the remark this answers is not in the document]
Comment (shwang) [reply src/imitation/data/types.py:153]: It's not immediately clear here from the implementation, but the reason this returns dictionary is because `Dict[str, Union[np.ndarray, Tensor]]` is a good `torch.util.data.Dataset` return type.

Pytorch natively knows how to join together lots of dictionaries to build dictionary batches, which is what we agreed would be good for `BC` training to use. `torch.util.data.DataLoader` samples shuffled batches from a `Transitions` instance (which recall, now implements `torch...Dataset`) by picking random indices, and then joining together the elements associated with each index.

I'll expand the comment here to explain that dictionary is easy for `DataLoader` to work with.

File: src/imitation/algorithms/bc.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (shwang) [reply src/imitation/algorithms/bc.py:52]: I went with `BCDataLoaderDucktype`

File: tests/test_data.py (hunk 3 above)
Comment (shwang) [reply tests/test_data.py:174]: That makes sense, I think that I just wanted group this close to the function that used it. Moving it to module level because it's more google-styley

File: src/imitation/data/types.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (shwang) [reply src/imitation/data/types.py:153]: I looked more closely into Torch's `default_collate` and it seems like they know how to join individual NamedTuple as well into batches as well.

However for the sake of allowing folks to easily use any ducktyped data source (without manufacturing a named tuple) it seems like a good idea to use dict with keys "obs" and "acts" in `BC`.

Comment (shwang) [inline src/imitation/data/types.py:153]: Torch's default collate doesn't play well with our `infos` attribute, as it tries to join together all the infos dicts into a single infos dict, rather than giving a list of dicts as we want... struggling to figure out a good solution to this problem right now.

I thought about hardcoding a "infos" case in `__getitem__` so that Torch sees a length-1 list of info dict whenever it is currently seeing a info dict, because joining together the length-1 lists without recursive joining will result in what we want. But this doesn't actually work, because Torch in fact will recursive join, and has an internal sanity check that errors if the info dicts have different numbers of keys on different samples: 
https://github.com/pytorch/pytorch/blob/c38a5cba0d4b310b0e24fbf6fe0fa14fe2b23d3a/torch/utils/data/_utils/collate.py#L76-L83

To fix this, I think that we will need to provide a custom collate_fn that performs the `infos` case first, and then defers to `default_collate` for everything else.

File: src/imitation/algorithms/bc.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (AdamGleave) [reply src/imitation/algorithms/bc.py:52]: This doesn't feel like ducktyping to me -- we're explicitly typing it. It's more an interface that Torch's `DataLoader` implements, but could be implemented by other classes. Perhaps something like `TorchDataset` would be more descriptive?

File: src/imitation/data/types.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (AdamGleave) [reply src/imitation/data/types.py:153]: OK, this seems reasonable, although I agree it's worth documenting the design rationale in a comment.

Comment (AdamGleave) [reply src/imitation/data/types.py:153]: Hmm seems like custom collate is the best way forward. Do we actually need `infos` for training though? I could also imagine some wrapper that just drops infos (or I think it's already optional in some places?)

[the remark this answers is not in the document]
Comment (shwang) [reply src/imitation/data/types.py:153]: Btw, now that I think about it, it's pretty trivial to construct arbitrary namedtuple on the fly, using the factory function: https://docs.python.org/3/library/collections.html#collections.namedtuple

[the remark this answers is not in the document]
Comment (shwang) [reply src/imitation/data/types.py:153]: ok, documented

File: src/imitation/algorithms/bc.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (shwang) [reply src/imitation/algorithms/bc.py:52]: `torch.util.data.Dataset` isalready a thing (`Transitions` implements that starting in this PR).

What about just `BCDataLoader`? (fwiw in the documentation, we will see `Iterable[Mapping[...` regardless of the alias name.

File: src/imitation/data/types.py (hunk 6 above)
Comment (shwang) [reply src/imitation/data/types.py:153]: The reason I added in `infos` in the first place was for a convoluted potential MineRL use where we might want to track which transitions belong to which trajectory name in our data stream.

If it turns out we don't need this in the future, I'd be fine with just removing `infos` from `TransitionsMInimal`.

> I think it's already otpional in some places?
It's a Optional field of `Trajectory`. However, it's required in `TransitionsMinimal` right now, even if it's just an empty dict for each timestep.

**Comment by shwang** [comment]
I won't merge this until after we have the analogous Adversarial dataset port PR ready.

I think it would be useful to know how easily the Pytorch Dataset abstractions work with Adversarial before the merge.

**Comment by shwang** [comment]
> This means we can no longer (easily) have algorithms with different batch sizes over time? Seems OK, a rather esoteric use case, but something to be aware of at least.

THis is still possible. We could just implement any iterator that returns a different number of samples in each batch, perhaps changing between different calls to `__iter__`.

**Comment by shwang** [comment]
[DataLoader interface](https://trello.com/c/2XqpcJUh/6-dataloader-interface)

**Comment by shwang** [comment]
Superceded by #236
